### PR TITLE
Remove unnecessary memory allocation limit checks and rename max-allocation related functions

### PR DESCRIPTION
**Pull request by micahsnyder**
The goal of the ClamAV's functions like "cli_malloc" and "cli_realloc" are to check that the size of the allocation is not unreasonable.  But we only need to do that check when the size is derived from untrusted user input (aka may vary based on values in the possibly malicious file that we're scanning). 

There are many calls to allocate memory where the size of the allocation is either known at compile time or else is derived from trusted user input.  Examples of variable but trusted allocation sizes include: 
- the number of parts of a signature
- or the number of blocks in a bytecode function. 

This PR seeks to eliminate those unnecessary max-allocation checks. This is mostly because checking the limits in these cases is a waste of CPU. But it also sets groundwork for a future improvement where we may pass in the memory allocation limit as a parameter, and make it generally configurable for the scanning service. This would be meaningful for anyone wishing to crank up the max-filesize to scan large archives that decompress to very large amounts of data. 

Further, this PR renames the max-allocation functions and related functions and macros so it's obvious what they do.  I suspect that the reason the likes of "cli_calloc" and "cli_realloc" were so widely used for fixed size allocations is that the developers knew we had a special wrapper for allocating memory and didn't know when they should be used.

What's also drawing attention to is the functions that wrap allocations but do NOT check the allocation-limit. These are `cli_strdup`, `cli_realloc`, and `cli_realloc2`.  These functions are safer than the original `strdup` and `realloc` functions from the C "standard library".  
- For `strdup`, our wrapper adds a NULL check to prevent accidental NULL-deref crashes.  
- For `realloc`, the "standard library" one is not standard. Some variations may free the pointer if size == 0. Some do not.  Ours wrapper does not.  In this way, we know the behavior will be consistent.  Our version also has two variants.  The first, `cli_realloc` does not free the pointer on failure.  The second, `cli_realloc2`, does free the pointer on failure.  That second behavior is generally less desirable because you must remember to set the pointer to NULL before `goto done;` or else you risk a double-free.  *I have not removed the latter variant simply because it's a lot of work, so there will still be two variants.*  In this PR, I improve the documentation and I rename those two so you can easily tell the difference.  

Also, for both the `strdup` and `realloc` wrappers, we now have a "cli_safer_" prefix on them when they do not check the the allocation limit.  When we do check the allocation limit, we instead use a "cli_max_" prefix.

Finally, for the macros which wrap the allocation-related functions, I've added a "_OR_GOTO_DONE" suffix to the macro names. This makes it clear what is special about the macros, and that error handling is built int.  If the allocation fails, then we `goto done;`. 

In summary, the functions and macros are now:
```c
void *cli_max_malloc(size_t nmemb);
void *cli_max_calloc(size_t nmemb, size_t size);
void *cli_max_realloc(void *ptr, size_t size);
void *cli_max_realloc_or_free(void *ptr, size_t size);
void *cli_safer_realloc(void *ptr, size_t size);
void *cli_safer_realloc_or_free(void *ptr, size_t size);
char *cli_safer_strdup(const char *s);

#define CLI_SAFER_STRDUP_OR_GOTO_DONE(buf, var, ...) \
...
#define CLI_FREE_AND_SET_NULL(var) \
...
#define CLI_MALLOC_OR_GOTO_DONE(var, size, ...) \
...
#define CLI_MAX_MALLOC_OR_GOTO_DONE(var, size, ...) \
...
#define CLI_CALLOC_OR_GOTO_DONE(var, nmemb, size, ...) \
...
#define CLI_MAX_CALLOC_OR_GOTO_DONE(var, nmemb, size, ...) \
...
#define CLI_VERIFY_POINTER_OR_GOTO_DONE(ptr, ...) \
...
#define CLI_MAX_REALLOC_OR_GOTO_DONE(ptr, size, ...) \
...
#define CLI_SAFER_REALLOC_OR_GOTO_DONE(ptr, size, ...) \
...
```

### CODE REVIEW

**Review by ragusaa — CHANGES_REQUESTED** [review]

File: clamonacc/inotif/inotif.c [hunk 1]
```diff
@@ -105,7 +105,7 @@ static int onas_ddd_init_wdlt(uint64_t nwatches)
 
     if (nwatches <= 0) return CL_EARG;
 
-    wdlt = (char **)cli_calloc(nwatches << 1, sizeof(char *));
+    wdlt = (char **)calloc(nwatches << 1, sizeof(char *));
```
[the remark this answers is not in the document]
Comment (micahsnyder) [reply clamonacc/inotif/inotif.c:108]: Per my note in the PR description, 

> we only need to do that check when the size is derived from untrusted user input (aka may vary based on values in the possibly malicious file that we're scanning).

System configuration, clamd/clamscan configuration, and signature databases are trusted input.  I'm aiming to only have allocation limits for allocation sizes derived from the file being scanned, or some API that an untrusted user could manipulate.

File: clamonacc/inotif/inotif.c [hunk 2]
```diff
@@ -243,7 +243,7 @@ static int onas_ddd_watch_hierarchy(const char *pathname, size_t len, int fd, ui
         curr = curr->next;
 
         size_t size      = len + strlen(curr->dirname) + 2;
-        char *child_path = (char *)cli_malloc(size);
+        char *child_path = (char *)malloc(size);
```
[the remark this answers is not in the document]
Comment (micahsnyder) [reply clamonacc/inotif/inotif.c:246]: Perhaps.  But practically speaking, our limit is like 1GB and the directory name is for an existing directory, and not arbitrary data.  That is, the attacker would have to make a directory with file path > 1GB to exceed our limit.  I'm not worried about it.

File: clamonacc/inotif/inotif.c [hunk 3]
```diff
@@ -330,7 +330,7 @@ static int onas_ddd_unwatch_hierarchy(const char *pathname, size_t len, int fd,
...
```
[the remark this answers is not in the document]
Comment (micahsnyder) [reply clamonacc/inotif/inotif.c:333]: Same as the other one.

File: clamonacc/inotif/inotif.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (ragusaa) [reply clamonacc/inotif/inotif.c:246]: ok

File: clamonacc/inotif/inotif.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (ragusaa) [reply clamonacc/inotif/inotif.c:108]: ok

File: clamonacc/inotif/inotif.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (ragusaa) [reply clamonacc/inotif/inotif.c:333]: ok

**Review by ragusaa — APPROVED** [review]

**Pull request merged**